unliftio/src/UnliftIO/Internals/Async.hs

```diff
@@ -835,3 +834,236 @@ mapConcurrently_ f t = withRunInIO $ \run -> A.mapConcurrently_ (run . f) t
+pooledMapConcurrently f xs = do
+  withRunInIO $ \run -> do
+    numProcs <- getNumCapabilities
+    pooledMapConcurrentlyIO numProcs (run . f) xs
+
+-- | Similar to 'pooledMapConcurrentlyN' but with flipped arguments.
+--
+-- @since 0.2.10
+pooledForConcurrentlyN :: (MonadUnliftIO m, Traversable t)
+                      => Int -- ^ Max. number of threads. Should not be less than 1.
+                      -> t a -> (a -> m b) -> m (t b)
+pooledForConcurrentlyN numProcs xs f = flip (pooledMapConcurrentlyN numProcs) xs f
+
+-- | Similar to 'pooledForConcurrentlyN' but with number of threads
+-- set from 'getNumCapabilities'. Usually this is useful for CPU bound
+-- tasks.
+--
+-- @since 0.2.10
+pooledForConcurrently :: (MonadUnliftIO m, Traversable t) => t a -> (a -> m b) -> m (t b)
+pooledForConcurrently xs f = flip pooledMapConcurrently xs f
+
+pooledMapConcurrentlyIO :: Traversable t => Int -> (a -> IO b) -> t a -> IO (t b)
+pooledMapConcurrentlyIO numProcs f xs =
+    if (numProcs < 1)
+    then error "pooledMapconcurrentlyIO: number of threads < 1"
+    else pooledMapConcurrentlyIO' numProcs f xs
+
+pooledConcurrently
+  :: Int -> MVar [a] -> (a -> IO b) -> IO ()
+pooledConcurrently numProcs jobsVar f = do
+  forConcurrently_ [1..numProcs] $ \_ -> do
+    let loop  = do
+          mbJob :: Maybe a <- modifyMVar jobsVar $ \x -> case x of
+            [] -> return ([], Nothing)
+            var : vars -> return (vars, Just var)
+          case mbJob of
+            Nothing -> return ()
+            Just x -> do
+              _ <- f x
+              loop
+    loop
+
+pooledMapConcurrentlyIO' ::
+  Traversable t => Int -> (a -> IO b) -> t a -> IO (t b)
+pooledMapConcurrentlyIO' numProcs f xs = do
+  -- prepare one IORef per result...
+  jobs :: t (a, IORef b) <-
+    for xs (\x -> (x, ) <$> newIORef (error "pooledMapConcurrentlyIO': empty IORef"))
+  -- ...put all the inputs in a queue..
+  jobsVar :: MVar [(a, IORef b)] <- newMVar (toList jobs)
+  -- ...run `numProcs` threads in parallel, each
+  -- of them consuming the queue and filling in
+  -- the respective IORefs.
+  pooledConcurrently numProcs jobsVar $ \ (x, outRef) -> f x >>= writeIORef outRef      -- Read all the IORefs
+  for jobs (\(_, outputRef) -> readIORef outputRef)
+
+pooledMapConcurrentlyIO_' ::
+  Traversable t => Int -> (a -> IO b) -> t a -> IO ()
+pooledMapConcurrentlyIO_' numProcs f jobs = do
+  jobsVar :: MVar [a] <- newMVar (toList jobs)
+  pooledConcurrently numProcs jobsVar f
+
+pooledMapConcurrentlyIO_ :: Traversable t => Int -> (a -> IO b) -> t a -> IO ()
+pooledMapConcurrentlyIO_ numProcs f xs =
+    if (numProcs < 1)
+    then error "pooledMapconcurrentlyIO_: number of threads < 1"
+    else pooledMapConcurrentlyIO_' numProcs f xs
+
+-- | Like 'pooledMapConcurrentlyN' but with the return value
+-- discarded.
+--
+-- @since 0.2.10
+pooledMapConcurrentlyN_ :: (MonadUnliftIO m, Traversable f)
```

Comment:
Can this be reduced to a `Foldable` constraint?

Reply:
Done. Also changed some other signatures to Foldable wherever it was possible.